### PR TITLE
Drop copy() reimplementation; use buffered I/O with larger buffers

### PR DESCRIPTION
**Pull request by bgilbert**
The current image pipeline is:

- local file or HTTP response →
- [some Readers that don't affect I/O size] →
- `BufReader` with 8 KiB buffer →
- decompressor →
- `copy()` with 256 KiB buffer →
- disk.

`copy()`'s large buffer doesn't matter here.  The decompressor will read data in whatever chunk size it wants, which will be aggregated only up to 8 KiB.  The decompressor will also produce data in whatever chunk size it wants, depending on the compressed stream, and those chunks won't be aggregated at all.

Upgrade the `BufReader` to a 256 KiB buffer and add 256 KiB `BufWriter`s.  This doesn't seem to affect performance much in my local tests, but it's still the right thing to do.  Switch back to `std::io::copy`, since our reimplementation was misguided; add buffers in notable places instead.

### CODE REVIEW

**Comment by cgwalters** [comment]
Sounds OK to me (didn't review code).

The way I'd measure things is comparing versus `curl | gunzip` basically.  

I do think we should investigate `O_DIRECT` and, maybe [using multiple threads](https://github.com/coreos/coreos-installer/issues/184#issuecomment-619240165) works better on top of this?

File: src/download.rs [hunk 1]
```diff
@@ -299,9 +296,13 @@ where
 pub fn image_copy_default(
     first_mb: &[u8],
     source: &mut dyn Read,
-    dest: &mut File,
+    dest_file: &mut File,
     _dest_path: &Path,
 ) -> Result<()> {
+    // Amortize write overhead.  The decompressor will produce bytes in
+    // whatever chunk size it chooses.
+    let mut dest = BufWriter::with_capacity(BUFFER_SIZE, dest_file);
```
Comment (lucab) [inline src/download.rs:304]: It would be better to flush this `BufWriter` before it gets dropped.

File: src/s390x/dasd.rs [hunk 2]
```diff
@@ -66,6 +66,9 @@ pub fn image_copy_s390x(
     // there shouldn't be any partition data in the first MiB, so don't
     // worry about copying first_mb
     let mut cursor: u64 = 1024 * 1024;
+    // amortize write overhead; the decompressor will produce bytes in
+    // whatever chunk size it chooses
+    let mut dest = BufWriter::with_capacity(BUFFER_SIZE, dest_file);
```
Comment (lucab) [inline src/s390x/dasd.rs:71]: It would be better to flush this `BufWriter` before it gets dropped.

**Comment by bgilbert** [comment]
Good catches.  Updated.

**Pull request merged**